test/test_collection.py

```diff
@@ -141,25 +158,106 @@ def test_watch(self):
+        self._compare_command_dicts(last_cmd_payload, last_logger_payload)
+
+    def test_find_one_and_delete(self):
+        logger = CommandLogger()
+        client = MongoClient(serverSelectionTimeoutMS=1000, event_listeners=[
+            logger])
+        collection = client.db.products
+        explain = ExplainCollection(collection)
+        collection.find_one_and_delete({"_id": "D"})
+        last_logger_payload = logger.cmd_payload
+        res = explain.find_one_and_delete({"_id": "D"})
+        self.assertIn("queryPlanner", res)
+        last_cmd_payload = explain.last_cmd_payload
+        self._compare_command_dicts(last_cmd_payload, last_logger_payload)
+
+    def test_find_one_and_replace(self):
+        logger = CommandLogger()
+        client = MongoClient(serverSelectionTimeoutMS=1000, event_listeners=[
+            logger])
+        collection = client.db.products
+        explain = ExplainCollection(collection)
+        collection.find_one_and_replace({'x': 1}, {'y': 1})
+        last_logger_payload = logger.cmd_payload
+        res = explain.find_one_and_replace({'x': 1}, {'y': 1})
+        self.assertIn("queryPlanner", res)
+        last_cmd_payload = explain.last_cmd_payload
+        self._compare_command_dicts(last_cmd_payload, last_logger_payload)
+
+    def test_find_one_and_update(self):
+        logger = CommandLogger()
+        client = MongoClient(serverSelectionTimeoutMS=1000, event_listeners=[
+            logger])
+        collection = client.db.products
+        explain = ExplainCollection(collection)
+        collection.find_one_and_update({'_id': 665}, {'$inc': {'count': 1}, "$set": {'done': True}})
+        last_logger_payload = logger.cmd_payload
+        res = explain.find_one_and_update({'_id': 665}, {'$inc': {'count': 1}, '$set': {'done': True}})
+        self.assertIn("queryPlanner", res)
+        last_cmd_payload = explain.last_cmd_payload
+        self._compare_command_dicts(last_cmd_payload, last_logger_payload)
+
+    def test_replace_one(self):
+        logger = CommandLogger()
+        client = MongoClient(serverSelectionTimeoutMS=1000, event_listeners=[
+            logger])
+        collection = client.db.products
+        explain = ExplainCollection(collection)
+        collection.replace_one({'x': 1}, {'y': 1},
+                               bypass_document_validation=True)
+        last_logger_payload = logger.cmd_payload
+        res = explain.replace_one({'x': 1}, {'y': 1}, bypass_document_validation=True)
+        self.assertIn("queryPlanner", res)
+        last_cmd_payload = explain.last_cmd_payload
+        self._compare_command_dicts(last_cmd_payload, last_logger_payload)
+
+    def test_estimated_document_count(self):
+        logger = CommandLogger()
+        client = MongoClient(serverSelectionTimeoutMS=1000, event_listeners=[
+            logger])
+        collection = client.db.products
+        explain = ExplainCollection(collection)
```

Comment:
It looks like we repeat this test setup code in every test:
```
        logger = CommandLogger()
        client = MongoClient(serverSelectionTimeoutMS=1000, event_listeners=[
            logger])
        collection = client.db.products
        explain = ExplainCollection(collection)
``` 

The unittest module has a feature dedicated to this kind of pattern which lets us avoid duplication. Please give this section a read: https://docs.python.org/3/library/unittest.html#organizing-test-code

We can also use the common test `setUp` method create the `client.db.products` collection (by inserting a document at the start of each test for example).

Reply:
Done.